dali/imgcodec/decoders/libtiff/tiff_libtiff.cc

```diff
@@ -223,21 +277,40 @@ DecodeResult LibTiffDecoderInstance::Decode(SampleView<CPUBackend> out, ImageSou
   TensorShape<> out_line_shape = {roi.shape()[1], out_channels};
   TensorShape<> out_line_strides = kernels::GetStrides(out_line_shape);
 
+  // We choose smallest possible type
+  size_t in_type_bits;
+  if (info.bit_depth <= 8) in_type_bits = 8;
+  else if (info.bit_depth <= 16) in_type_bits = 16;
+  else if (info.bit_depth <= 32) in_type_bits = 32;
+  else
+    DALI_FAIL(make_string("Unsupported bit depth: ", info.bit_depth));
+
   TYPE_SWITCH(out.type(), type2id, OutType, (IMGCODEC_TYPES), (
-    VALUE_SWITCH(info.bit_depth, Depth, (8, 16, 32), (
-      using InputType = detail::depth2type<Depth>::type;
-      auto *row_in  = static_cast<InputType *>(row_buf.get());
+    VALUE_SWITCH(in_type_bits, InTypeBits, (8, 16, 32), (
+      using InputType = detail::depth2type<InTypeBits>::type;
+
+      std::vector<InputType> unpacked;
+      InputType *row_in;
+      if (info.bit_depth == InTypeBits) {
+        row_in = static_cast<InputType *>(row_buf.get());
+      } else {
+        unpacked.resize(volume(out_line_shape));
+        row_in = unpacked.data();
+      }
+
       OutType *const img_out = out.mutable_data<OutType>();
-      for (int64_t y = 0; y < roi.shape()[0]; y++) {
-        LIBTIFF_CALL(TIFFReadScanline(tiff.get(), row_in, roi.begin[0] + y, 0));
-        Convert(img_out + (y * out_row_stride), out_line_strides.data(), 1, opts.format,
+      for (int64_t roi_y = 0; roi_y < roi.shape()[0]; roi_y++) {
+        LIBTIFF_CALL(TIFFReadScanline(tiff.get(), row_buf.get(), roi.begin[0] + roi_y, 0));
+        if (info.bit_depth != InTypeBits) {
+          detail::UnpackBits(info.bit_depth, row_in, row_buf.get(), volume(out_line_shape));
```

Comment:
Trying to understand. This is not part of convert because in TIFF the bits are packed, so it's not simply scaling to the dynamic range of the type. Correct?

Reply:
Yes, that's correct. I need to do some bitwise magic to unpack them before giving them to Convert